src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/HeaderFileBuilder.java

```diff
@@ -89,6 +92,11 @@ public void addFunction(String javaName, String nativeName, FunctionInfo functio
             Constant mhConstant = constantBuilder.addMethodHandle(javaName, nativeName, functionInfo, false)
                     .emitGetter(this, MEMBER_MODS, Constant.QUALIFIED_NAME, nativeName);
             emitFunctionWrapper(mhConstant, javaName, nativeName, functionInfo);
+            if (functionInfo.methodType().returnType().equals(MemorySegment.class)) {
+                // emit scoped overload
+                emitFunctionWrapperNoAllocatorOverload(javaName, functionInfo);
+                emitFunctionWrapperScopedOverload(javaName, functionInfo);
+            }
```

Comment:
With the emitFunctionWrapper call above this, don't we get a publicly accessible function wrapper that will always throw a WrongMethodTypeException because it's missing an allocator argument? Maybe it should be conditionally made private? (since the other 2 functions reference it, it's still needed).

Reply:
I see what you mean.

Reply:
Actually, I think the code is correct, as the generated "main" method will either have no allocator argument (in case return is not a struct) or will have allocator parameter (in case return is struct). The issue with the code is that, when no  allocator argument is present, we should pass DEFAULT_ALLOCATOR to the underlying MH.

Reply:
Scratch that - we don't need a new allocator parameter when struct is not returned as the underlying MH doesn't take an allocator.

I think the code is fine as is.

Reply:
Ah, right. I see now we get:
```
MS func(SegmentAllocator, ...)
MS func(...) // uses default scope
MS func(ResourceScope, ...)
```

I misread this is the first overload not taking an allocator. What you have looks good.